Note changes in bytecode between versions

rocky opened the pull request:
I'm done with a first pass. Since this is large, short of whatever small things we can catch, I'd like this merged in even though there are gaps in filling out newly added instructions, and other things of the "finish completing..." nature. For that, I think we just make more passes.

I'll be branching off of this branch for the other couple of ideas I have. (And that's another reason why I'd like this back into master - I'd rather branch from there.)

rocky committed:
WIP: Start noting changes in bytecode.

(BTW Stefan was dead wrong about backward compatiblity and no drastic semantics changes in ops. Look at opcode 97 which is alternatively "mark", "mult", and "save-current-buffer".)

rocky committed:
WIP: Fill in more version differences

rocky committed:
Add version 22.1 - 23.1 changes (none)

rocky committed:
Update changes up to version 24.1...

Some stack-ref0->stack-ref[0] kinds of changes. More is needed.

stack-ref example added. Note what's up with BYTE_CODE_SAFE.

pipcet commented on elisp-bytecode.texi:
```diff
@@ -537,7 +538,38 @@ opcode, in little-endian byte order.
+Running an instruction with opcode 0 (logically this would be called
+@code{stack-ref[0]}), will cause an immediate abort of Emacs!
+
+0 is typically an invalid bytecode and assembly instructions because 0
```
It's an invalid opcode, but a common and valid byte in a bytecode string. I think that's worth pointing out.

rocky replied:
If you have specific wording, let me know. I'm not sure what would change.

Turns out in 25.1 the behviour change prints an error and doesn't abort. I'll push the updates to that soon.

pipcet replied:
I'd just change "invalid bytecode and assembly instructions" to "invalid opcode".

pipcet commented on elisp-bytecode.texi:
```diff
@@ -1029,7 +1093,10 @@ Call @code{+} with two arguments, adding the two values at the top of the stack.
+Call @code{*} with two arguments, multiplying the two values at the
+top of the stack.
+
+Since emacs 19.34
```
bytecomp.el incorrectly states "v19 only" for this opcode, but it's still emitted.

pipcet commented on elisp-bytecode.texi:
```diff
@@ -1063,6 +1132,7 @@ function for them.
 * concat2::
 * concat3::
 * concat4::
+* concatN::
```
I had deliberately excluded concatN and listN from this section, but I think it's better to document them here rather than having a special section for them, now.

pipcet commented on elisp-bytecode.texi:
```diff
@@ -1090,26 +1160,44 @@ Call @code{concat} with three arguments.
+@unnumberedsubsubsec @code{concatN} (174)
+@kindex concatN
+
+Call @code{concat} an arbitrary number of arguments.  Note there are
```
in the 0-255 range, concat with more arguments will generate a function call.

rocky replied:
That makes more sense of the generated code. Have fixed and in the opcode table it will be pushed as soon as I go over the rest.

pipcet commented on elisp-bytecode.texi:
```diff
@@ -1224,76 +1323,128 @@ Call @code{current-buffer} with no arguments.
 @kindex set-buffer
 Call @code{set-buffer} with one argument.
 
+@node save-current-buffer-1
```
I don't think we should call this "save-current-buffer-1", since "byte-save-current-buffer" is what bytecomp.el uses.

rocky replied:
Ok change, and will push soon.

pipcet commented on elisp-bytecode.texi:
```diff
@@ -1224,76 +1323,128 @@ Call @code{current-buffer} with no arguments.
+@node save-current-buffer-1
+@unnumberedsubsubsec @code{save-current-buffer-1} (114)
+@kindex save-current-buffer
+Call @code{save-current-buffer} with one stack argument,
```
There is no stack argument, as far as I can see.

rocky replied:
Thanks. I was wondering what was going on in the bytecode.

pipcet commented on elisp-bytecode.texi:
```diff
@@ -1331,9 +1482,145 @@ new binding when executed.  They need to be balanced with
+intruction was never generated.
+
+Starting in this version, unless C prepocessor variable
+@code{BYTE_CODE_SAFE} (on by default) is defined, the obsolete
```
off by default, I think.

pipcet commented on elisp-bytecode.texi:
```diff
@@ -1331,9 +1482,145 @@ new binding when executed.  They need to be balanced with
+
+Is replaced by @code{save-current-buffer-1} (114)
+
+Unless C prepocessor variable @code{BYTE_CODE_SAFE} (on by default) is
```
ditto

rocky replied:
Ok. Changed and will go in next push.

pipcet commented on elisp-bytecode.texi:
```diff
@@ -1878,62 +2185,79 @@ Make a binding recording buffer, point, and mark.
+@item @verb{|0162|}
+@tab @verb{| 114|}
+@tab @verb{|save-current-buffer-1|}
+@tab Call @code{save-current-buffer} with one argument.
```
Again, I don't think this is best described as a "call" instruction.

rocky replied:
Ok. Suggest something.

pipcet commented on elisp-bytecode.texi:
```diff
@@ -1878,62 +2185,79 @@ Make a binding recording buffer, point, and mark.
 @item @verb{|0161|}
 @tab @verb{| 113|}
 @tab @verb{|  set-buffer|}
-@tab Call @code{set-buffer} with one argument.
```
Why was this removed? Is it incorrect?

rocky replied:
Probably an inadvertent mistake. Have readded and will push soon.

rocky replied:
Addressed in 8a3903d7 . Not sure why this still shows. Also, I think we should remove the "with one argument" stuff as those are indicated in the stack effects. In this table, I think we want just the minimum and the full description in the Instructions section.  I would like a link with no extra verbiage which is why I started the other index. However I haven't been able to pull that off yet.

I think the stack effects might also appear the instruction section. However, I'll creating a couple more branches to try out some new ideas: 

* Breaking up the opcode table into a couple of sequential tables
* Better descriptions in the Instruction section.

pipcet replied:
I agree on th "with one argument" thing, it'd be best to omit both that and the descriptions and just write "call *" or "call +"

pipcet commented on elisp-bytecode.texi:
```diff
@@ -896,6 +935,16 @@ Call @code{list} with three arguments.
+@node listN
+@unnumberedsubsubsec @code{list4} (70)
+@kindex listN
+Call @code{list} an arbitrary number of arguments.
```
Again, "up to 255".

rocky replied:
Thanks. Have fixed the xxxN, and will push when I am finished with the comments.

pipcet reviewed (commented):
Looks good overall! I'll be travelling for a few days, FWIW.

rocky committed:
Fold pipcet's comments in...

Remove in opcode table some of the "with one argument", "with two arguments"
as that is indicated in the stack effect column.

rocky committed:
Add a couple more bytecode examples

rocky committed:
Add bytecode headers...

Interestingly, useful information has been dropped from the header.

rocky committed:
Add stack-set ops

pipcet commented on elisp-bytecode.texi:
```diff
@@ -896,6 +935,14 @@ Call @code{list} with three arguments.
 Call @code{list} with four arguments.
 
+@node listN
+@unnumberedsubsubsec @code{list4} (70)
```
`listN` here

rocky committed:
cut-n-paste typo list4->listN

rocky committed:
Finish changes up to Emacs 27.1...

Note switch instruction

pipcet commented on elisp-bytecode.texi:
```diff
@@ -1133,6 +1244,7 @@ but simply call the corresponding C function.
 * bobp::
 * current-buffer::
 * set-buffer::
+* byte-save-current-buffer::
```
I'd mildly prefer `save-current-buffer-OBSOLETE` and `save-current-buffer` as our names for these instructions.

pipcet commented on elisp-bytecode.texi:
```diff
@@ -1222,90 +1342,165 @@ Call @code{current-buffer} with no arguments.
+@code{(set-buffer TOS)}
+
+@strong{Example}
+@code{(defun set-buffer-eg() (set-buffer "*scratch"))} generates:
```
I'd use "`*scratch*`" in this example.

pipcet commented on elisp-bytecode.texi:
```diff
@@ -1222,90 +1342,165 @@ Call @code{current-buffer} with no arguments.
+@node discardN
+@unnumberedsubsec @code{discardN} (180)
+@kindex discardN
+Discards up to 255 arguments from the stack.  Note there is a special
```
argument -> value, I'd say.

pipcet commented on elisp-bytecode.texi:
```diff
@@ -1320,6 +1515,22 @@ PC  Byte  Instruction
+@node stack-set2
+@unnumberedsubsec @code{stack-setN} (181)
+@kindex stack-set2
+Like discardN. [What's the difference?]
```
I don't think that's true. stack-set discards only one value, after copying it to a position further up the stack; however, stack-set[0] is indeed equivalent to discard.

pipcet commented on elisp-bytecode.texi:
```diff
@@ -1320,6 +1515,22 @@ PC  Byte  Instruction
+@kindex stack-set2
+Like discardN. [What's the difference?]
+Can handle up to 255 arguments.  Note there is a special
+instruction when there is only one argument.
```
stack-set handles up to 255 values, stack-set2 handles up to 65535 values.

pipcet commented on elisp-bytecode.texi:
```diff
@@ -1331,9 +1542,272 @@ new binding when executed.  They need to be balanced with
+The bytecode meta-comment no longer includess the bytecomp version used.
+
+Bytecode meta-comments look like this:
+@verbatim
```
what about the `;ELC^W^@^@^@` line? Is that part of the bytecode meta-comment?

pipcet commented on elisp-bytecode.texi:
```diff
@@ -1344,164 +1818,164 @@ Make a binding recording buffer, point, and mark.
 @item @verb{|  06|}
 @tab @verb{|   6|}
-@tab @verb{|  stack-ref6|}
+@tab @verb{|  stack-ref[6]|}
```
I think I messed up disassembling this instruction correctly, but I'll have to check. Should it be, say, `stack-ref[6] [255]`

pipcet commented on elisp-bytecode.texi:
```diff
@@ -2244,7 +2828,7 @@ Make a binding recording buffer, point, and mark.
 @tab @verb{| 214|}
 @tab @verb{|  constant-i22|}
-@tab
+     @tab
```
I don't understand this change, was it intentional?

pipcet reviewed (approved):
Let's merge this if it's any help at all, it's definitely a step in the right direction.
